src/core/lwip/tcp_out.c

```diff
@@ -1885,6 +1885,44 @@ void tcp_rexmit_rto(struct tcp_pcb *pcb)
     if (pcb->unacked == NULL) {
         return;
     }
+    /*  Before moving all unacked to unsent, we need to maintain correct order of segments.
+        For exmaple, in unsent we might find a retransmitted segments:
+        unacked: 1->2->3
+        unsent: 4->5->6
+
+        After fast retransmission of seg 1,2:
+        unacked: 3
+        unsent: 1->2->4->5->6
+
+        After RTO without handling correct order:
+        unacked: ----
+        unsent: 3->1->2->4->5->6
+
+        Therefore - we first move back retransmitted segments to unacked:
+        unacked: 1->2->3
+        unsent: 4->5->6
+        And then move all unacked to unsent:
+        unacked: ----
+        unsent: 1->2->3->4->5->6
+    */
+
+    if (pcb->unsent != NULL && TCP_SEQ_GT(pcb->unacked->seqno, pcb->unsent->seqno)) {
+        // Move retransmitted segments to unacked
+        struct tcp_seg *orig_unsent = pcb->unsent;
+        struct tcp_seg *last_unsent_move_to_unacked = pcb->unsent;
```

Comment:
Long variable names can complicate code. For short code blocks it's better to keep short names .

Reply:
Done.